corehq/blobs/migrate.py

```diff
@@ -0,0 +1,241 @@
+import os
+import traceback
+from cStringIO import StringIO
+from tempfile import mkdtemp
+
+from django.conf import settings
+from couchexport.models import SavedBasicExport
+from corehq.blobs import get_blob_db
+from corehq.blobs.models import BlobMigrationState
+from corehq.dbaccessors.couchapps.all_docs import (
+    get_all_docs_with_doc_types,
+    get_doc_count_by_type,
+)
+from couchdbkit import ResourceConflict
+
+MIGRATION_INSTRUCTIONS = """
+There are {total} documents that may have attachments, and they must be
+migrated to a new blob database.
+
+Run these commands to procede with migrations:
+
+./manage.py run_blob_migration {slug} --file=FILE
+./manage.py migrate
+
+Note: --file=FILE is optional and can be omitted if you do not want to
+keep a copy of the pre-migrated couch documents.
+
+See also:
+https://github.com/dimagi/commcare-hq/blob/master/corehq/blobs/migrate.py
+"""
+
+BLOB_DB_NOT_CONFIGURED = """
+Cannot get the blob database.
+
+This often means that settings.SHARED_DRIVE_ROOT is not configured.
+It should be set to a real directory. Update localsettings.py and
+retry the migration.
+"""
+
+MIGRATIONS_SKIPPED_WARNING = """
+WARNING {} documents were not migrated due to concurrent modification
+during migration. Run the migration again until you do not see this
+message.
+"""
+
+
+class Migrator(object):
+
+    def __init__(self, slug, doc_types):
+        self.slug = slug
+        self.doc_types = doc_types
+
+    def migrate(self, filename=None):
+        return migrate(self.slug, self.doc_types, filename)
+
+
+MIGRATIONS = {m.slug: m for m in [
+    Migrator("saved_exports", [SavedBasicExport]),
+]}
+
+
+def migrate(slug, doc_types, filename=None):
+    """Migrate attachments from couchdb to blob storage
+
+    The blob storage backend is associated with the model class.
+
+    :param doc_types: List of couch model classes with attachments to be migrated.
+    :param filename: File path for intermediate storage of migration data.
+    :returns: A tuple `(<docs migrated>, <docs skipped>)`
+    """
+    couchdb = doc_types[0].get_db()
+    assert all(t.get_db() is couchdb for t in doc_types[1:]), repr(doc_types)
+    type_map = {cls.__name__: cls for cls in doc_types}
+
+    dirpath = None
+    if filename is None:
+        dirpath = mkdtemp()
+        filename = os.path.join(dirpath, "export.txt")
+
+    print("Loading documents: {}...".format(", ".join(type_map)))
+    total = 0
+    skips = 0
+    with open(filename, 'w') as f:
+        for doc in get_all_docs_with_doc_types(couchdb, list(type_map)):
+            if doc.get("_attachments"):
+                f.write('{}\n'.format(json.dumps(doc)))
+                total += 1
+
+    with open(filename, 'r') as f:
+        for n, line in enumerate(f):
+            if n % 100 == 0:
+                print_status(n + 1, total)
+            doc = json.loads(line)
+            obj = type_map[doc["doc_type"]](doc)
+            try:
+                with obj.atomic_blobs():
```

Comment:
No longer use `IterDB` because `atomic_blobs()` saves the document to couch on `__exit__`. This eliminates the problem with blobs being written and then orphaned if the object cannot be saved. Hopefully this won't be too detrimental to performance.

Question: when `obj.save()` results in a `ResourceConflict` error, is the document in couch left in a conflicted state?